snippets/core/system-text-json/csharp/ImmutablePointConverter.cs

```diff
@@ -27,42 +41,35 @@ public class ImmutablePointConverter : JsonConverter<ImmutablePoint>
             bool ySet = false;
 
             // Get the first property.
+            ReadXorY(ref reader, options, ref x, ref xSet, ref y, ref ySet);
+
+            // Get the second property.
+            ReadXorY(ref reader, options, ref x, ref xSet, ref y, ref ySet);
+
             reader.Read();
-            if (reader.TokenType != JsonTokenType.PropertyName)
-            {
-                throw new JsonException();
-            }
 
-            string propertyName = reader.GetString();
-            if (propertyName == XName)
-            {
-                x = ReadProperty(ref reader, options);
-                xSet = true;
-            }
-            else if (propertyName == YName)
-            {
-                y = ReadProperty(ref reader, options);
-                ySet = true;
-            }
-            else
+            if (reader.TokenType != JsonTokenType.EndObject)
             {
                 throw new JsonException();
             }
 
-            // Get the second property.
+            return new ImmutablePoint(x, y);
+        }
+
+        private void ReadXorY(ref Utf8JsonReader reader, JsonSerializerOptions options, ref int x, ref bool xSet, ref int y, ref bool ySet)
```

Comment:
I believe this is incorrect.

The following would succeed, incorrectly: `{"X": 1, "X": 1}`

The second call to `ReadXorY` needs different conditions (from https://github.com/dotnet/samples/pull/1837#discussion_r361062959). Note the `xSet` and `ySet` being used.
```C#
        if (xSet && reader.ValueTextEquals(YName.EncodedUtf8Bytes))
        {
            y = ReadProperty(ref reader, options);
        }
        else if (ySet && reader.ValueTextEquals(XName.EncodedUtf8Bytes))
        {
            x = ReadProperty(ref reader, options);
        }
        else
        {
            throw new JsonException();
        }
```

In this case, I would remove the `ReadXorY` method and just inline the logic (like you had before). Passing the bools by ref also makes it hard to reason about. The two "if blocks" are not exactly identical in my sample, so it doesn't help to move it into a helper.